janus_core/processing/observables.py

```diff
@@ -1,72 +1,307 @@
 
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING
+
 from ase import Atoms, units
 
+if TYPE_CHECKING:
+    from janus_core.helpers.janus_types import SliceLike
+    from janus_core.helpers.utils import slicelike_len_for
+
+
+# pylint: disable=too-few-public-methods
+class Observable(ABC):
+    """
+    Observable data that may be correlated.
+
+    Parameters
+    ----------
+    dimension : int
+        The dimension of the observed data.
+    """
+
+    def __init__(self, dimension: int = 1):
+        """
+        Initialise an observable with a given dimensionality.
+
+        Parameters
+        ----------
+        dimension : int
+            The dimension of the observed data.
+        """
+        self._dimension = dimension
+
+    @abstractmethod
+    def __call__(self, atoms: Atoms) -> list[float]:
+        """
+        Signature for returning observed value from atoms.
+
+        Parameters
+        ----------
+        atoms : Atoms
+            Atoms object to extract values from.
+
+        Returns
+        -------
+        list[float]
+            The observed value, with dimensions atoms by self.dimension.
+        """
+
+    def value_count(self, n_atoms: int | None = None) -> int:
+        """
+        Count of values returned by __call__.
+
+        Parameters
+        ----------
+        n_atoms : int | None
+            Atom count to expand atoms_slice.
+
+        Returns
+        -------
+        int
+            The number of values returned by __call__.
+        """
+        return self.dimension
+
+    @property
+    def dimension(self):
+        """
+        Dimension of the observable. Commensurate with self.__call__.
+
+        Returns
+        -------
+        int
+            Observables dimension.
+        """
+        return self._dimension
+
+
+class ComponentMixin:
+    """
+    Mixin to handle Observables with components.
+
+    Parameters
+    ----------
+    components : dict[str, int]
+        Symbolic components mapped to indices.
+    """
+
+    def __init__(self, components: dict[str, int]):
+        """
+        Initialise the mixin with components.
+
+        Parameters
+        ----------
+        components : dict[str, int]
+            Symbolic components mapped to indices.
+        """
+        self._components = components
+
+    @property
+    def allowed_components(self) -> dict[str, int]:
+        """
+        Allowed symbolic components with associated indices.
+
+        Returns
+        -------
+        dict[str, int]
+            The allowed components and associated indices.
+        """
+        return self._components
+
+    @property
+    def _indices(self) -> list[int]:
+        """
+        Get indices associated with self._components.
+
+        Returns
+        -------
+        list[int]
+            The indices for each self._components.
+        """
+        return [self._components[c] for c in self.components]
```

Comment:
Is this just `list(self._components.values())`?

Reply:
this may lead to duplicate indices. e.g. stress supports both "xy" and "yx" as ```self._components```. ```self.components``` are the particular (say stress) components an observable "tracks" hence the need to access those particular values from the dict.

I've changed this so there is no ```allowed_components``` property, just a ```_allowed_components``` member to track this. Hopefully makes this clearer.